internal/jobtracker/jobtracker.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2025 Canonical.
+
+// Package jobtracker provides a way to run routines in one instance of JIMM and track them in another.
+// That is, their status can be checked and they can be stopped.
+
+package jobtracker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/juju/clock"
+	"github.com/juju/retry"
+	"github.com/juju/zaputil/zapctx"
+	"go.uber.org/zap"
+)
+
+// JobTrackerStore defines the interface for tracking the lifecycle and status of jobs.
+// It provides methods to add a new job, update its status (running, successful, or failed),
+// and retrieve a stop signal for a specific job.
+type JobTrackerStore interface {
+	AddJob(ctx context.Context, jobType string) (uuid.UUID, error)
+	SetJobRunning(ctx context.Context, jobId uuid.UUID) error
+	SetJobSuccessful(ctx context.Context, jobId uuid.UUID) error
+	SetJobFailed(ctx context.Context, jobId uuid.UUID, jobErr error) error
+	GetJobStopSignal(ctx context.Context, jobId uuid.UUID) (stopSignal bool, err error)
+}
+
+// Tracker manages job tracking operations using a provided JobTrackerStore.
+// It periodically performs tasks based on the specified stopInterval duration.
+type Tracker struct {
+	store        JobTrackerStore
+	stopInterval time.Duration
+}
+
+// NewJobTracker creates and returns a new Tracker instance using the provided JobTrackerStore and stopInterval.
+// It returns an error if the store is nil or if stopInterval is not greater than zero.
+func NewJobTracker(store JobTrackerStore, stopInterval time.Duration) (*Tracker, error) {
+	tracker := &Tracker{}
+	if store == nil {
+		return tracker, errors.New("store cannot be nil")
+	}
+	if stopInterval <= 0 {
+		return tracker, errors.New("stopInterval must be greater than zero")
+	}
+	tracker.stopInterval = stopInterval
+	tracker.store = store
+
+	return tracker, nil
+}
+
+// Run runs a new job and reurns the job ID.
+func (j *Tracker) Run(ctx context.Context, jobType string, job func(ctx context.Context) error, deadline time.Duration) (uuid.UUID, error) {
+	jobId, err := j.store.AddJob(ctx, jobType)
+	if err != nil {
+		return jobId, err
+	}
+
+	go j.manageJob(ctx, jobId, deadline, job)
+
+	return jobId, nil
+}
+
+// manageJob runs a job with a given context, job ID, polling interval, and deadline.
+// It manages the job's lifecycle, including setting its status in the store, handling retries on store operations,
+// and responding to stop signals or context cancellations. The job is run in a separate goroutine, and its status
+// is updated as running, successful, or failed based on its result or context expiration.
+// If a stop signal is received or the context is canceled or times out, the job is marked as failed.
+// Store operations are retried up to 5 times with a 30-second delay between attempts in case of transient errors.
+func (j *Tracker) manageJob(
```

Comment:
`handleJob` or `startJob` read better to me than "manage". Also not opposed to `runJob`.

Reply:
or simply `do`